src/Compilers/CSharp/Portable/Binder/DecisionDagBuilder_ListPatterns.cs

```diff
@@ -0,0 +1,66 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using Microsoft.CodeAnalysis.CSharp.Symbols;
+using Microsoft.CodeAnalysis.PooledObjects;
+
+namespace Microsoft.CodeAnalysis.CSharp
+{
+    internal sealed partial class DecisionDagBuilder
+    {
+        private void MakeTestsAndBindingsForLengthAndListPatterns(BoundDagTemp input, BoundRecursivePattern recursive, ArrayBuilder<BoundPatternBinding> bindings, ArrayBuilder<Tests> tests)
+        {
+            if (recursive.HasErrors)
+                return;
+
+            Debug.Assert(recursive.LengthProperty is not null || input.Type.IsSZArray());
+            var syntax = recursive.Syntax;
+            var lengthProperty = recursive.LengthProperty ?? (PropertySymbol)_compilation.GetSpecialTypeMember(SpecialMember.System_Array__Length);
+            var lengthEvaluation = new BoundDagPropertyEvaluation(syntax, lengthProperty, input);
+            tests.Add(new Tests.One(lengthEvaluation));
```

Comment:
Based on the recent LDM, I'd expect that `array is { .. }` wouldn't actually evaluate the length property. It's the things _other_ than `..` that cause `Length` to be evaluated.

Reply:
We'll add tests explicitly for such optimizations in a later PR (I'm preparing such tests and they are behaving well already). Those optimizations happen later in the pipeline.

Reply:
From the notes 

>We think that .. without a trailing pattern and without any other element patterns is similar in concept to a _ pattern

It looks like we do actually want this to behave like that. @jcouv can you confirm?

Reply:
Opened https://github.com/dotnet/csharplang/pull/4790 to update the spec.

Reply:
> emit only a null test

Yes, `{..}` should just do a null test, because of the braces.
Sorry if it's pedantic, in my mind, `..` itself does not include a null test. That's why we can drop the `Slice` call in many scenarios (we don't have any checks to perform on what the value it would return anyways).

Reply:
The null test is inherited from the containing recursive pattern. If `{}` checks for null, `{..}` should too.

Reply:
Oh I don't mean to test the return value, only the original input to the recursive pattern.